botocore/signers.py

```diff
@@ -130,3 +148,320 @@ def get_auth(self, signing_name, region_name, signature_version=None):
+
+            {'url': 'https://mybucket.s3.amazonaws.com
+             'fields': {'acl': 'public-read',
+                        'key': 'mykey',
+                        'signature': 'mysignature',
+                        'policy': 'mybase64 encoded policy'}
+            }
+        """
+        if fields is None:
+            fields = {}
+
+        if conditions is None:
+            conditions = []
+
+        if region_name is None:
+            region_name = self._request_signer.region_name
+
+        # Create the policy for the post.
+        policy = {}
+
+        # Create an expiration date for the policy
+        datetime_now = datetime.datetime.utcnow()
+        expire_date = datetime_now + datetime.timedelta(seconds=expires_in)
+        policy['expiration'] = expire_date.strftime(botocore.auth.ISO8601)
+
+        # Append all of the conditions that the user supplied.
+        policy['conditions'] = []
+        for condition in conditions:
+            policy['conditions'].append(condition)
+
+        # Obtain the appropriate signer.
+        query_prefix = '-presign-post'
+        signature_version = self._request_signer.signature_version
+        if not signature_version.endswith(query_prefix):
+            signature_version += query_prefix
+
+        kwargs = {'signing_name': self._request_signer.signing_name,
+                  'region_name': region_name,
+                  'signature_version': signature_version}
+
+        signature_type = signature_version.split('-', 1)[0]
+
+        try:
+            auth = self._request_signer.get_auth(**kwargs)
+        except UnknownSignatureVersionError:
+            raise UnsupportedSignatureVersionError(
+                signature_version=signature_type)
+
+        # Store the policy and the fields in the request for signing
+        request = create_request_object(request_dict)
+        request.context['s3-presign-post-fields'] = fields
+        request.context['s3-presign-post-policy'] = policy
+
+        auth.add_auth(request)
+
+        # Fix s3 host for s3 sigv2 bucket names
+        fix_s3_host(request, signature_type, region_name)
+        # Return the url and the fields for th form to post.
+        return {'url': request.url, 'fields': fields}
+
+
+def add_generate_presigned_url(class_attributes, **kwargs):
+    class_attributes['generate_presigned_url'] = generate_presigned_url
+
+
+def generate_presigned_url(self, ClientMethod, Params=None, ExpiresIn=3600,
+                           HTTPMethod=None):
+    """Generate a presigned url given a client, its method, and arguments
+
+    :type ClientMethod: string
+    :param ClientMethod: The client method to presign for
+
+    :type Params: dict
+    :param Params: The parameters normally passed to
+        ``ClientMethod``.
+
+    :type ExpiresIn: int
+    :param expires_in: The number of seconds the presigned url is valid
+        for. By default it expires in an hour (3600 seconds)
+
+    :type HTTPMethod: string
```

Comment:
Yeah that is fine with me. I do not have to big a preference for one of the other. I did it the way I did because acronyms are usually capitalized.